### ISSUE TITLE
Debug not works with "Sam local start-api" with "--debug-port" and "--warm-containers"

### ISSUE THREAD

**Issue by CeccoCQ** [aws/aws-sam-cli#3749]
With this command:
`sam local start-api --template-file template.yml --region eu-west-1 --debug-port 5555 --warm-containers EAGER`

and this launch.json
```
{
      "name": "Attach to SAM CLI",
      "type": "node",
      "request": "attach",
      "address": "localhost",
      "port": 5555,
      "localRoot": "${fileDirname}",
      "remoteRoot": "/var/task",
      "protocol": "inspector",
      "stopOnEntry": false
    }
```

my Visual Studio Code doesn't attach to container, I have no error or log, simply after 5 seconds the debug's topbar disappear.
The terminal logs doesn't show "Debugging listening on...." when I try lo execute an HTTP request.

All works fine removing --warm-containers  param.

**Comment by Cliftonz** [comment]
Getting the same issue.

**Comment by jaedenharris** [comment]
I also cannot start the debugger mode with or without the --warm-containers option

`sam local start-api --warm-containers LAZY -p 3001 -d 9999`

**Comment by achandoff** [comment]
Am getting the same issue. any update ?

**Comment by ibadsiddiqui** [comment]
Hello Everyone,
I am having the same issue, the debugger doesn't attach and the top bar disappears.

**Comment by vfilter** [comment]
Debug does not work at all for me. Using the built-in scaffolded hello world and executing `sam local invoke -d 9999 -e events/event.json HelloWorldFunction` will successfully invoke the Lambda, but won't wait for the debugger. The hello world function just gets executed and ends.

**Comment by jonkipu** [comment]
@[CeccoCQ](https://github.com/CeccoCQ) Regarding the original issue using `start-api`, I had to dive into the code a bit to find out the following:

If the template contains several functions (could be generated functions besides your original lambda, such as log retention related lambdas), a warning will be output: 
```Warning: you supplied debugging options but you did not specify the --debug-function option.  To specify which function you want to debug, please use the --debug-function <function-name>```

What's non-intuitive about this code path is that it will disable setting of the debug port rendering debugging useless in this case. If you have one function and don't specifiy `--debug-function` AWS SAM will assume you meant to debug your one and only function, when using ```---warm-containers``` the code path will disable debugging if the function being executed isn't the one specified by ```--debug-function```, however the code uses different identifiers of said function to check, again rendering debugging useless, this honestly seems like a bug.

To make sure you can debug your function use ```--debug-function``` with your current flags, but specify the function name as specified in your template, e.g: `MyFunction1A65A8F` as opposed to just `MyFunction`

**Comment by CeccoCQ** [comment]
@jonkipu thanks for your answer.
I've a lot of Lambda Functions and for my purposes I've to debug them concurrently, I mean more function at the time. 
With ---debug-function the debug will be restricted only to one function, I suppose that this is a bug or if it's a feature, it's a really strange feature (in my opinion) because if I've an API gateway at 99% I've several functions.

**Comment by jonkipu** [comment]
I agree, I was baffled by this myself, it seems that for whatever reason, when no ```--debug-function``` is passed, debugging is disabled.
I have yet to try and change the code to see if it just works when enabling debugging and debugging multiple functions, if it does I'll update you and I guess we'll have to bring this issue into someone's attention

**Comment by jonkipu** [comment]
@[CeccoCQ](https://github.com/CeccoCQ), played with the code a little and I think part of this is by design, since when opting to use ```--warm-containers```, SAM will retain the containers between invocations of each function, since it uses a different container for each function under your API, it can only expose the debug options (namely the ```--debug-port```) to the host OS for only one of those containers, that's why choosing ```--debug-function``` is required, since it controls for which of your functions those debug settings are applied.
If you forego ```--warm-containers``` SAM will create a container per function but will destroy them after that function is finished

**Comment by OrRosenblatt** [comment]
I run into the same issue. When running either of the following:
- `sam local start-api --template-file template.yml --debug-port 5858 --warm-containers LAZY`
- `sam local start-api --template-file template.yml --debug-port 5858 --warm-containers EAGER` 

The specific ENV variables related to debugging (`NODE_OPTIONS` , `NODE_PATH` etc.) weren't added to the Docker image, and the 5858 debug port wasn't exposed. I also see the warning message [mentioned above](https://github.com/aws/aws-sam-cli/issues/3749#issuecomment-1184205969):

```txt
Warning: you supplied debugging options but you did not specify the --debug-function option.  
To specify which function you want to debug, please use the --debug-function <function-name>.
```

However when running `sam local start-api --template-file template.yml --debug-port 5858` the port and the related node options are setup correctly on the Docker image. Any update on this issue?

**Comment by jonkipu** [comment]
@OrRosenblatt not sure this is so easy to fix / will be fixed soon as mentioned in my comment.
I've been using the --debug-function parameter when using "--warm-containers" since then and it's fine as a workaround (at least for me)

**Comment by vikyathshirva** [comment]
is there an update on this? making debug work for multiple functions would help :)

**Comment by Peter-Schorn** [comment]
I ran into what appears to be the same symptom with `sam local start-api --debug-port 5858 --warm-containers EAGER`, and in my case the important detail was that the debugger port was not actually usable until the Lambda had been invoked at least once.

With `EAGER`, SAM reported that the container was already running and ready, but VS Code still could not attach. The sequence looked like this:

```text
Containers created. Waiting for readiness...
All containers are ready.
Running on http://127.0.0.1:3000
```

At this point, attaching to port 5858 failed.

Then I sent the first HTTP request to the local API. That request initially hung, and only then did SAM print:

```text
Invoking dist/index.handler (nodejs24.x)
START RequestId: ...
Debugger listening on ws://0.0.0.0:5858/...
```

After that, attaching from VS Code succeeded immediately, the waiting request resumed, breakpoints worked, and subsequent requests could be debugged normally.

So, for anyone who lands here with a warm container that is supposedly “ready” but VS Code says it cannot connect to a debuggable target, try this exact order:

1. Start `sam local start-api` with `--debug-port`.
2. **Invoke the function once** (for example, send an HTTP request to the local API endpoint).
3. The first request may hang while waiting for the debugger.
4. Now attach VS Code to the debug port.
5. Once attached, the request continues and breakpoints work.

In other words, at least for Node.js, `--warm-containers EAGER` appears to create/start the container eagerly, but it does not start the Node inspector eagerly. The inspector only begins listening when an actual invocation starts.

This is separate from the `--debug-function` / multiple-functions discussion above. I reproduced it with a single Lambda function. It would be helpful if SAM either started the inspector as part of eager container initialization, or at least documented/logged that the first invocation is required before the debug target exists.
